disk_objectstore/utils.py

```diff
@@ -1519,3 +1534,117 @@ def merge_sorted(iterator1: Iterable[Any], iterator2: Iterable[Any]) -> Iterator
+        #  > already provide multithreaded implementations.
+        # Therefore, in the future if we support those clever algorithms we can just skip this and let the algorithm
+        # do its job.
+
+        # Estimate the amount of compression we can get and decide whether to compress based on how much we can save.
+        compression_ratio = estimate_compression(source_stream, source_size)
+        return compression_ratio < compression_threshold
+    raise NotImplementedError(f"Unknown {compress_mode=}")
+
+
+def estimate_compression(stream: StreamSeekBytesType, size: int) -> float:
+    """Tries to quickly estimate the compression ratio of the file.
+
+    At the end the stream, put back at the same position as when we entered this method (i.e. we will seek back
+    at the stream.tell() position that was set at the function enter).
+
+    :param stream: a seekable stream.
+    :param size: the total size of the stream (this is used to try to get small chunks of data from small
+        random parts of the stream.
+    :return: the ratio between the expected compressed size and the total file size. E.g. 1. means compressing
+        does not provide any feedback, 0.5 means compression halves the files size, 0.1 means the compressed version
+        only occupies 10% of the space, a value larger than 1 means that compressing actually increases the file size,
+        etc. Note that this is only a quick estimate; actual compression levels might vary, but it is meant as a
+        quick heuristics to decide whether we should compress or not, when CompressMode.AUTO is requested.
+    """
+    # Set some internal parameters
+    sample_size = 1024  # Get this amount of bytes...
+    # Only check a sampled_data of at most `max_sample_data_size` bytes (approximately)
+    max_sampled_data_size = 131_072  # 128kB
+    # Approximately, except for rounding errors, get a small sample of size `sample_size` every so many bytes
+    sample_interval = size // (max_sampled_data_size // sample_size)
+
+    # Get the current pos (to go back to the same position at the end)
+    initial_pos = stream.tell()
+    # Go back to the beginning
+    stream.seek(0)
+
+    sampled_data = b""
+    # Only check the first part of the file
+    while len(sampled_data) < max_sampled_data_size:
+        chunk = stream.read(sample_size)
+        if not chunk:
+            # EOF, stop
+            break
+        sampled_data += chunk
+        # Advance to the next sample of length SAMPLE_INTERVAL
+        # Never go back!
+        stream.seek(max(0, sample_interval - len(chunk)), 1)
+    # I now have the sample, I compress it and compare the size
+
+    # I just want to know if it's compressible, I compress with some cheap ZLIB version,
+    # Ideally this gives me a good estimate. Probably I could even do better, e.g. like BTRFS I
+    # could estimate Shannon's entropy
```

Comment:
I think the 2 and 3 test https://github.com/aiidateam/disk-objectstore/issues/14 here would still be good and fast to test if a file is compressible, without going for the Shannon's entropy. 
They basically check the number of unique bytes in the sample, which can be used to detect if the file is a text file, which is likely to benefit. 

But I think the current approach is good enough, as we are not aiming for maximizing performance.

Reply:
Another note, GZIP uses a sliding window of 32 KB to "cache" the content, so if we use a 1 KB sample length the compression ratio can be a bit overestimated.

Reply:
OK, I will use 32kB probably then. However, I won't use the number of characters - if I understand correctly they check if there are less than 64 different characters - I tried, but in most "real" files (e.g. python scripts etc) I have more like 80-90 characters. We can improve in the future.

Reply:
Maybe 32KB is a bit too large - I am not entirely sure. We can use something in between perhaps.